perf/light.cc

```diff
@@ -0,0 +1,651 @@
+		TupleHolder data(state.range(0));
+		state.ResumeTiming();
+
+		auto lookup_iter = data.tuples.rbegin();
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			benchmark::DoNotOptimize(light_find(&ht,
+				lookup_iter->hash, lookup_iter->raw));
+			lookup_iter++;
+			processed += 2;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, InsertAndFind)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+// Delete random values till the hashtable is empty.
+BENCHMARK_DEFINE_F(LightFixture, DeleteRandValue)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_delete_value(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, DeleteRandValue)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+////////////////////////////// Unordered set Fixture ///////////////////////////////////////////////////////////////////
+
+struct TupleEqual
+{
+	bool operator()(const Tuple &lhs, const Tuple &rhs) const noexcept
+	{
+		return std::memcmp(lhs.raw.data, rhs.raw.data, lhs.raw.size) == 0;
+	}
+};
+
+// Just to compare with Light. These measurements are not assumed to be the
+// most closest to ones related to Light - hash tables interfaces seem to be
+// too different. It's just about making Light numbers a bit representative.
+class UnorderedSetFixture : public benchmark::Fixture {
+public:
+	void
+	TearDown(const benchmark::State &state) noexcept final
+	{
+		set.clear();
+	}
+
+protected:
+	void
+	Fill(std::vector<Tuple>::iterator &&begin, std::vector<Tuple>::iterator &&end) noexcept
+	{
+		set.insert(begin, end);
+	}
+
+	void
+	Erase(std::vector<Tuple>::iterator &&begin, std::vector<Tuple>::iterator &&end) noexcept
+	{
+		std::for_each(begin, end,
+			[this](const Tuple &t){ set.erase(t);});
+	}
+
+	std::unordered_set<Tuple, Hash::TupleHash, TupleEqual> set;
+};
+
+BENCHMARK_DEFINE_F(UnorderedSetFixture, InsertRandValue)
+(benchmark::State& state)
+{
+	std::size_t insertion_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(set.insert(v));
+			insertion_count++;
+		}
+	}
+	state.SetItemsProcessed(insertion_count);
+}
+
+BENCHMARK_REGISTER_F(UnorderedSetFixture, InsertRandValue)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+// Light can't reserve buckets beforehand, so let's benchmark inserts with
+// reserve separately. Apparently inserts with reserve turn out to be slower.
```

Comment:
Perhaps this comment will be wrong after fixes I suggested.

Reply:
Fixed.